### PR TITLE
[WIP] Improve Tunable and parameter optimization

### PR DESCRIPTION
**Pull request by kiudee**
This starts the work on #26. Input on the implementation is welcome.

Goal
----
The new parameter optimization should be able to remove as much algorithm dependent code for handling hyperparameters as possible.
To this end we transfer all the handling of ranges and conversions of
points to dictionaries to the parameter optimizer.
The interface `Tunable` is also simplified. Thus each tunable algorithm only has to implement `set_tunable_parameters` because it knows best how to apply changes to hyperparameters.

TODO
----

- [x] Change Tunable to new interface
- [x] Let ParameterOptimizer handle ranges and tunables
- [x] Adjust FATE ranker
- [x] Adjust FETA ranker
- [x] Adjust misc rankers
- [x] Simplify ParameterOptimizer class
- [x] Test ParameterOptimizer class

### CODE REVIEW

**Comment by prithagupta** [comment]
Completes all parts except test. Can you please review it

File: csrank/fate_ranking.py [hunk 1]
```diff
@@ -593,55 +537,15 @@ def predict_scores(self, X, **kwargs):
...
-        for k, v in dict_params.items():
-            cls._tunable[k] = v
-        cls._tunable = OrderedDict(sorted(cls._tunable.items()))
-
-    @classmethod
-    def set_tunable_parameter_ranges(cls, param_ranges_dict):
-        logger = logging.getLogger(GENERAL_OBJECT_RANKING_CORE)
-        return tunable_parameters_ranges(cls, logger, param_ranges_dict)
+    def set_tunable_parameters(self, n_hidden_set_units=32, n_hidden_set_layers=2, **point):
+        FATERankingCore.set_tunable_parameters(self, **point)
```
Comment (kiudee) [inline csrank/fate_ranking.py]: We should also offer the parameters of the superclass here.
Otherwise the users do not know that there are more parameters like `n_hidden_joint_layers`.

File: csrank/fate_ranking.py [hunk 2]
```diff
@@ -382,22 +333,12 @@ def _fit(self, X=None, Y=None, generator=None, epochs=35, inner_epochs=1,
...
-            for c in callbacks:
-                if isinstance(c, LRScheduler):
-                    c.initial_lr = K.get_value(self.optimizer.lr)
-                    self.logger.info("Setting lr {} for {}".format(c.initial_lr, c.__name__))
-            if self._use_early_stopping:
-                callbacks.append(self.early_stopping)
-
-            self.logger.info("Callbacks {}".format(
-                ', '.join([c.__name__ for c in callbacks])))
+                for c in callbacks:
```
Comment (kiudee) [inline csrank/fate_ranking.py]: `callbacks` can be None, which will fail with an error (see test).

File: csrank/tests/test_tuning.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+@fixture
+def optimizer():
+    from ..tunable import Tunable
+
+    class RankerStub(Tunable):
+        def __init__(self):
+            pass
+
+        def set_tunable_parameters(self, point):
+            self.__dict__.update(point)
```
Comment (kiudee) [inline csrank/tests/test_tuning.py]: Needs to be changed to the new `Tunable` interface

File: csrank/tuning.py [hunk 4]
```diff
@@ -41,59 +41,57 @@ def on_iteration_end(self, i, logs=None):
...
+        optimizer_path :
+        tunable_parameter_ranges :
+        ranker_params :
+        fit_params :
+        random_state :
+        tuning_callbacks :
+        validation_loss :
+        learning_problem :
+        kwd :
+        """
```
[the remark this answers is not in the document]
Comment (prithagupta) [reply csrank/tuning.py]: Will complete this.

File: csrank/objectranking/object_ranker.py [hunk 5]
```diff
@@ -141,11 +142,10 @@ def __call__(self, X, *args, **kwargs):
         """
         return self.predict(X, **kwargs)
 
-    def set_init_lr_callback(self, callbacks):
+    def set_initial_lr_for_scheduler(self, callbacks):
         for c in callbacks:
             if isinstance(c, LRScheduler):
                 c.initial_lr = K.get_value(self.optimizer.lr)
-                self.logger.info("Setting lr {} for {}".format(c.initial_lr, c.__name__))
         return callbacks
```
[the remark this answers is not in the document]
Comment (prithagupta) [reply csrank/objectranking/object_ranker.py]: But what if the user is not using this callback at all?

File: csrank/callbacks.py [hunk 6]
```diff
@@ -94,9 +103,16 @@ def step_decay(self, epoch):
         lrate = self.initial_lr * math.pow(self.drop, step)
         return lrate
 
+    def set_tunable_parameters(self, epochs_drop=300, drop=0.1, **point):
```
[the remark this answers is not in the document]
Comment (prithagupta) [reply csrank/callbacks.py]: No but that is the one which is set for the Gradient Descent algorithm.

File: csrank/fate_ranking.py (hunk 1 above)
Comment (prithagupta) [reply csrank/fate_ranking.py]: done

File: csrank/objectranking/object_ranker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kiudee) [reply csrank/objectranking/object_ranker.py]: Then there is no problem. The learning rate of the optimizer can also be tuned for the rankers separately.

[the remark this answers is not in the document]
Comment (prithagupta) [reply csrank/objectranking/object_ranker.py]: Then we will have two separate Lr values if we are using two lr as tunable and also even if the user is not providing the optimizer lr as tunable for the the optimizer it will still set the learning rate initial value to the optimizer learning rate

[the remark this answers is not in the document]
Comment (prithagupta) [reply csrank/objectranking/object_ranker.py]: https://github.com/keras-team/keras/blob/ef13db05731bfd53fa0a877637c99c1734be933b/keras/callbacks.py#L626
We can simply remove initial_lr as the new keras API is able to handle it by itself for setting the lr for the callback same as that for the optimizer
